runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/interceptors/Md5ChecksumInterceptor.kt

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.http.interceptors
+
+import aws.smithy.kotlin.runtime.client.ProtocolRequestInterceptorContext
+import aws.smithy.kotlin.runtime.hashing.md5
+import aws.smithy.kotlin.runtime.http.HttpBody
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.http.request.header
+import aws.smithy.kotlin.runtime.http.request.toBuilder
+import aws.smithy.kotlin.runtime.util.InternalApi
+import aws.smithy.kotlin.runtime.util.encodeBase64String
+
+/**
+ * Set the `Content-MD5` header based on the current payload
+ * See:
+ *   - https://awslabs.github.io/smithy/1.0/spec/core/behavior-traits.html#httpchecksumrequired-trait
+ *   - https://datatracker.ietf.org/doc/html/rfc1864.html
+ */
+@InternalApi
+public class Md5ChecksumInterceptor<I>(
+    private val block: ((input: I) -> Boolean)? = null,
+) : HttpInterceptor {
+
+    private var shouldInjectMD5Header: Boolean = false
+
+    override fun readAfterSerialization(context: ProtocolRequestInterceptorContext<Any, HttpRequest>) {
+        shouldInjectMD5Header = block?.let {
+            @Suppress("UNCHECKED_CAST")
+            val input = context.request as I
+            it(input)
+        } ?: true
+    }
+
+    override suspend fun modifyBeforeRetryLoop(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        if (!shouldInjectMD5Header) {
+            return context.protocolRequest
+        }
+
+        val checksum = when (val body = context.protocolRequest.body) {
+            is HttpBody.Bytes -> body.bytes().md5().encodeBase64String()
```

Comment:
correctness: this doesn't handle anything other than in-memory http bodies, is that correct?

Reply:
The way it's implemented today only handles in-memory byte bodies, there is even a [test case specifically for that behavior](https://github.com/awslabs/smithy-kotlin/blob/3326e72797d868258cbb84a6aedb686fe306cb3a/runtime/protocol/http/common/test/aws/smithy/kotlin/runtime/http/middleware/Md5ChecksumTest.kt#L56-L70). 

Is it possible we would need to handle other types of bodies here? I'm not sure about the context around the original implementation of this and why it was only set to work on `HttpBody.Bytes`